cmd/rook/ceph/osd.go

```diff
@@ -418,35 +416,54 @@ func readCephSecret(path string) error {
 	return nil
 }
 
-func destroyOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) (*oposd.OSDInfo, error) {
-	osdInfo, err := osddaemon.GetOSDInfoById(context, clusterInfo, osdID)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get OSD info for OSD.%d", osdID)
+// destroyOSD fetches the OSD to be replaced based on the ID and then destroys that OSD and zaps the backing disk
+func destroyOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, id int) (*osd.OSDReplaceInfo, error) {
+	var osd *osd.OSDReplaceInfo
+	var err error
+	if cfg.pvcBacked {
+		osd = &oposd.OSDReplaceInfo{ID: id, Path: fmt.Sprintf("/mnt/%s", os.Getenv(oposd.PVCNameEnvVarName))}
+	} else {
+		osd, err = osddaemon.GetOSDReplaceInfoById(context, clusterInfo, id)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get OSD info for OSD.%d", id)
+		}
 	}
 
-	// destroy the osd
-	logger.Infof("destroying OSD %d on path %q in %q mode", osdInfo.ID, osdInfo.BlockPath, osdInfo.CVMode)
-	destroyOSDArgs := []string{"osd", "destroy", fmt.Sprintf("osd.%d", osdInfo.ID), "--yes-i-really-mean-it"}
+	logger.Infof("destroying OSD %d", osd.ID)
+	destroyOSDArgs := []string{"osd", "destroy", fmt.Sprintf("osd.%d", osd.ID), "--yes-i-really-mean-it"}
 	_, err = client.NewCephCommand(context, clusterInfo, destroyOSDArgs).Run()
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to destroy osd.%d.", osdInfo.ID)
+		return nil, errors.Wrapf(err, "failed to destroy osd.%d.", osd.ID)
 	}
+	logger.Infof("successfully destroyed osd.%d", osd.ID)
 
-	// Sanitize OSD disk
-	s := cleanup.NewDiskSanitizer(context, clusterInfo,
-		&cephv1.SanitizeDisksSpec{
-			Method:     cephv1.SanitizeMethodProperty(cephv1.SanitizeMethodComplete),
-			DataSource: cephv1.SanitizeDataSourceProperty(cephv1.SanitizeDataSourceZero),
-			Iteration:  1,
-		},
-	)
+	// remove the dm device
+	if cfg.pvcBacked && cfg.storeConfig.EncryptedDevice {
+		target := oposd.EncryptionDMName(os.Getenv(oposd.PVCNameEnvVarName), oposd.DmcryptBlockType)
+		err = osddaemon.RemoveEncryptedDevice(context, target)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to remove dm device %q", target)
+		}
+	}
 
-	// TODO: handle disk sanitization errors
-	if osdInfo.CVMode == "raw" {
-		s.SanitizeRawDisk([]oposd.OSDInfo{osdInfo})
-	} else if osdInfo.CVMode == "lvm" {
-		s.SanitizeLVMDisk([]oposd.OSDInfo{osdInfo})
+	// zap the device
+	output, err := context.Executor.ExecuteCommandWithCombinedOutput("stdbuf", "-oL", "ceph-volume", "lvm", "zap", osd.Path, "--destroy")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to zap osd.%d path %q. %s.", osd.ID, osd.Path, output)
 	}
 
-	return &osdInfo, nil
+	logger.Infof("%s\n", output)
+	logger.Infof("successfully zaped osd.%d path %q", osd.ID, osd.Path)
+
+	// shred the device?
+	shredArgs := []string{"--random-source=/dev/zero", "--force", "--verbose", "--iterations=1", osd.Path}
```

Comment:
Can we limit the shred to just a small amount (e.g. 1MB) so we don't have to wait for the full disk?

Reply:
Ceph-Volume team recommended not to use `shred` and that `ceph-volume lvm zap` should be enough. It works for both `raw` and `lvm` mode.  I've removed the `shred` call for now.  Tested it and looks like `ceph-volume zap` should be enough.